Enable RSA-PSS padding in EVP_Digest* API

julek-wolfssl opened the pull request.

julek-wolfssl committed:
Enable RSA-PSS padding in EVP_Digest* API

bgermann reviewed (changes requested):
OpenSSL's default salt length is not RSA_PSS_SALT_LEN_DEFAULT but RSA_PSS_SALT_LEN_DISCOVER. As long as EVP_PKEY_CTX_set_rsa_pss_saltlen is not implemented, that one should be used.

dgarske commented on src/ssl.c:
```diff
@@ -30672,6 +30640,43 @@ int wolfSSL_RSA_sign_ex(int type, const unsigned char* m,
+        WOLFSSL_MSG("Bad function arguments");
+        return 0;
+    }
+    show("Message to Sign", m, mLen);
```
Can you refactor this "show" macro... it is a bad macro name. Hadn't seen it until this. How about `DEBUG_SIGN()`?

dgarske commented on src/ssl.c:
```diff
@@ -30714,32 +30721,71 @@ int wolfSSL_RSA_sign_ex(int type, const unsigned char* m,
+            case RSA_NO_PADDING:
+            case RSA_PKCS1_PSS_PADDING:
+            case RSA_PKCS1_OAEP_PADDING:
+                ret = SSL_SUCCESS;
```
Let's use the `WOLFSSL_SUCCESS` enum value internally.

julek-wolfssl replied:
I changed all `SSL_SUCCESS` to `WOLFSSL_SUCCESS`.

dgarske commented on src/ssl.c:
```diff
@@ -30758,65 +30804,87 @@ int wolfSSL_RSA_sign_ex(int type, const unsigned char* m,
+    return wolfSSL_RSA_verify_ex(type, m, mLen, sig, sigLen, rsa, RSA_PKCS1_PADDING);
+}
+
+#define wolfSSL_RSA_verify_ex_return(msg, ret_code) {               \
```
Macros for cleanup seem problematic. I'd prefer a goto end over this.

julek-wolfssl replied:
Todd asked I keep goto's to a minimum but in this case I guess it makes more sense.

dgarske replied:
There are ways to refactor the logic to not use a macro or goto. That flow is usually:

```c
/* proper init of all stack varibles / args */

ret = funa();
if (ret == 0)
   ret = funcb();
if (ret == 0)
   ret = funcc();

/* all caleanup here */
```

julek-wolfssl replied:
I used a goto because it is simpler with functions that have different success conditions.

dgarske commented on src/ssl.c:
```diff
@@ -30672,6 +30640,43 @@ int wolfSSL_RSA_sign_ex(int type, const unsigned char* m,
+            RSA_PKCS1_PADDING);
+}
+
+int wolfSSL_RSA_sign_ex(int type, const unsigned char* m,
```
Please add documentation either doxy style above or in the "doc/dox_comments/header_files/rsa.h". For example the "int flag" argument is not clear.

dgarske commented on wolfcrypt/src/evp.c:
```diff
@@ -2394,8 +2394,8 @@ int wolfSSL_EVP_DigestSignFinal(WOLFSSL_EVP_MD_CTX *ctx, unsigned char *sig,
                 break;
-            ret = wolfSSL_RSA_sign(nid, digest, hashLen, sig, &sigSz,
-                                   ctx->pctx->pkey->rsa);
+            ret = wolfSSL_RSA_sign_ex(nid, digest, hashLen, sig, &sigSz,
```
What does "flag = 1" mean / do here?

julek-wolfssl replied:
`flag = 1` means that `wolfSSL_RSA_sign_ex` only returns the value to which the return value of `*Verify*` should be compared to. In effect this returns the `wc_EncodeSignature` output for `RSA_PKCS1_PADDING` and the input for all other paddings.

dgarske replied:
Thanks, please document this in the code or in the doxy

dgarske commented on wolfcrypt/src/rsa.c:
```diff
@@ -3197,11 +3236,12 @@ int wc_RsaSSL_VerifyInline(byte* in, word32 inLen, byte** out, RsaKey* key)
 
 int  wc_RsaSSL_Verify_ex(const byte* in, word32 inLen, byte* out, word32 outLen,
-                         RsaKey* key, int pad_type)
+                         RsaKey* key, int pad_type, enum wc_HashType hash)
```
This will break public API compatibility for `wc_RsaSSL_Verify_ex`. I realize it was added fairly recently, but I believe it was in v4.5.0 release... and we've had support requests for it.. Also no documentation for it.

julek-wolfssl replied:
Should I change it to something else like `wc_RsaSSL_Verify_padding`?

dgarske replied:
I dislike continually adding new ex type API's. Consider something more broad if you need to expand this. Also consider there might be another way using an internal API.

julek-wolfssl replied:
An additional concern: having `wc_RsaSSL_Verify_ex` with the `pad_type` parameter but without a way to pass in the chosen hash is limiting. If a user wants `WC_RSA_PKCSV15_PAD` padding then they already can use the `wc_RsaSSL_Verify` API. I would argue that the hash parameter is a necessary addition for this API.

mouse07410 replied:
How does OpenSSL do it? What stops you from doing the same?

julek-wolfssl replied:
> How does OpenSSL do it? What stops you from doing the same?

OpenSSL internals have a very different design to wolfSSL. Doing the same is not ideal since the compatibility layer is meant to provide an interface to wolfSSL native functions. We want to keep the compatibility layer as clean as possible.

dgarske replied:
Keep compatibility with openssl. You know best.

julek-wolfssl replied:
`wc_RsaSSL_Verify_ex` was left unchanged. I added `wc_RsaSSL_Verify_ex2` which has the necessary parameters for the compatibility layer.

julek-wolfssl commented:
Thanks @bgermann for checking. I'll look into salt length and what the value should be.

julek-wolfssl committed:
enum wc_HashType switch 

switch needs to handle all possible enum values or else the compiler generates warnings

julek-wolfssl committed:
Refactoring and use salt length discover if available

julek-wolfssl committed:
CAVP, Windows, and FIPS tests

dgarske commented on wolfcrypt/src/rsa.c:
```diff
@@ -1736,6 +1736,60 @@ int wc_RsaUnPad_ex(byte* pkcsBlock, word32 pkcsBlockLen, byte** out,
+#ifndef NO_SHA
+        return WC_MGF1SHA1;
+#else
+        WOLFSSL_MSG("Unrecognized or unsupported hash function");
```
Can you refactor this to eliminate the duplicate MSG and return code logic? Use break and single unrecognized exit. Also maybe rename this function to include wc_Rsa since its a "local" function?

dgarske commented on wolfssl/openssl/rsa.h:
```diff
@@ -116,10 +116,14 @@ WOLFSSL_API int wolfSSL_RSA_sign(int type, const unsigned char* m,
+WOLFSSL_API int wolfSSL_RSA_verify_ex(int type, const unsigned char* m,
+                               unsigned int mLen, const unsigned char* sig,
+                               unsigned int sigLen, WOLFSSL_RSA* rsa,
+                               int padding);
```
I am still concerned about breaking compatibility with these _ex API's, since it was in the last v4.5.0 release. I know we've seen support tickets about the use of these API's, so I don't think customers will be happy. You may have to add another version of the API's perhaps even more generic to cover all future cases (you can use _ex2 if you want). Also the documentation for this API is missing.

julek-wolfssl replied:
Could it be called `wolfSSL_RSA_verify_generic_padding` for example?

dgarske commented on src/ssl.c:
```diff
@@ -30672,6 +30640,62 @@ int wolfSSL_RSA_sign_ex(int type, const unsigned char* m,
+    }
+    DEBUG_SIGN_msg("Message to Sign", m, mLen);
 
     if (rsa->inSet == 0)
```
Please cleanup the braces in this function. They are inconsistent and if with macro like `WOLFSSL_MSG` should have braces.

Examples:

```
if (outLen == 0)
        WOLFSSL_MSG("Bad RSA size");
```

```
if (ret == WOLFSSL_SUCCESS)
        WOLFSSL_MSG("wolfSSL_RSA_sign success");
    else {
        WOLFSSL_MSG("wolfSSL_RSA_sign failed");
    }
```

dgarske commented on src/ssl.c:
```diff
@@ -30672,6 +30640,62 @@ int wolfSSL_RSA_sign_ex(int type, const unsigned char* m,
+ * @param padding   Padding to use. Only RSA_PKCS1_PSS_PADDING and
+ *                  RSA_PKCS1_PADDING are currently supported for signing.
+ * @return          WOLFSSL_SUCCESS on success and WOLFSSL_FAILURE on error
+ */
```
This needs added into the doxy rsa.h.

julek-wolfssl replied:
The rsa.h doxy only includes the wolfCrypt RSA functions. Should I add it anyway?

dgarske replied:
Put in the appropriate location... ssl.h

dgarske commented on src/ssl.c:
```diff
@@ -30672,6 +30640,62 @@ int wolfSSL_RSA_sign_ex(int type, const unsigned char* m,
+
+    if (m == NULL || sigRet == NULL || sigLen == NULL || rsa == NULL) {
+        WOLFSSL_MSG("Bad function arguments");
+        return 0;
```
Please cleanup these "failure" `return 0` to use `WOLFSSL_FAILURE`. There are more below.

julek-wolfssl committed:
Review changes

dgarske reviewed (approved):
Thanks @julek-wolfssl , great work on this.

julek-wolfssl commented:
Thanks @dgarske !
